src/coprocessor/dag/expr/builtin_like.rs

```diff
@@ -33,26 +32,20 @@ impl ScalarFunc {
     }
 
     pub fn regexp(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
-        let target = try_opt!(self.children[0].eval_string(ctx, row));
-        let pattern = try_opt!(self.children[1].eval_string(ctx, row));
-
-        let mut p = String::from("(?i)");
-        p.push_str(str::from_utf8(&pattern)?);
-        let t = str::from_utf8(&target)?;
+        let target = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let pattern = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
+        let pattern = format!("(?i){}", &pattern);
 
         // TODO: cache compiled result
-        Ok(Some(Regex::new(&p)?.is_match(t) as i64))
+        Ok(Some(Regex::new(&pattern)?.is_match(&target) as i64))
     }
 
     pub fn regexp_binary(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
         let target = try_opt!(self.children[0].eval_string(ctx, row));
-        let pattern = try_opt!(self.children[1].eval_string(ctx, row));
-
-        let p = str::from_utf8(&pattern)?;
-        let t = str::from_utf8(&target)?;
+        let pattern = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
 
         // TODO: cache compiled result
-        Ok(Some(Regex::new(p)?.is_match(t) as i64))
+        Ok(Some(BytesRegex::new(&pattern)?.is_match(&target) as i64))
```

Comment:
Seems `target` should be evaluated by `eval_string_and_decode` too?

Reply:
In `regexp` yes, but `regexp_binary` not because the `target` could be an invalid `utf8` bytes.